### DISCUSSION TITLE
Automate releases and generate CHANGELOG in Angular Projects

### DISCUSSION THREAD

**Discussion by tomavic** [Zhengqbbb/cz-git#51] [General]
Installing `cz-git` is one step of many steps being added to Angular projects to protect your commit history from random bad commit messages and keep it clean and convenient so that we can generate a nice `CHANGELOG.md` for each release we generate. Starting from this point, where you need to automate your releases using to generate a semantic releases. 

At this point we have two options: use [Standard Version](https://github.com/conventional-changelog/standard-version) which is deprecated (I don't know why 🤯) or use [Semantic Release](https://semantic-release.gitbook.io/semantic-release/).

Before following the [Setup](https://semantic-release.gitbook.io/semantic-release/usage/installation) and [Configuration](https://semantic-release.gitbook.io/semantic-release/usage/configuration) of Semantic Release, I need to prepare my Angular project with the following:

- Add Husky hooks to run before commit and push.
- Keep the beauty of the code using Prettier, ESLint and Stylelint
- Install lint staged to run linting on staged files only at the time of pre-commit hook.
- Run Unit test or build scripts before pushing (Optional Extra step).

After finishing the previous steps, Here comes the part of generating a release.

What is the best changelog packages to be installed along with `cz`

**Comment by Zhengqbbb** [comment]
Do you want to generate CHANGELOG.md or github release page?

### Generate CHANGELOG.md

I use pkgs `bumpp@7.2.0` and `conventional-changelog-cli`

Reference Vuepress repository: [https://github.com/vuepress/vuepress-next/blob/392297fa856fd5869de40e4999c5bc4d126a3941/package.json#L22-L26](https://github.com/vuepress/vuepress-next/blob/392297fa856fd5869de40e4999c5bc4d126a3941/package.json#L22-L26)

But I am running the `Shell` script: [https://github.com/Zhengqbbb/cz-git/blob/e2b402d59be36da6a4a5a1ae446978850f10986e/.x-cmd/deploy#L31-L35](https://github.com/Zhengqbbb/cz-git/blob/e2b402d59be36da6a4a5a1ae446978850f10986e/.x-cmd/deploy#L31-L35)

### Github release page record changlog

I recommend using github action: [https://github.com/antfu/changelogithub](https://github.com/antfu/changelogithub)

### How to work

Use the repo`git log` command to get the commit message stream for two version number parsing

### Why use `bumpp@7.2.0` and `conventional-changelog-cli`
- Small, Single Feature, easy to split
- `bumpp`: change the version number for package.json, auto commit and auto tag and push
- Also `bumpp` **can insert additional commands to run after the version number is changed**
- At this point, we can use `conventional-changelog-cli` to add changelog

**Reply by tomavic** [reply]
The problem with conventional changlog and standard version that is deprecated

**Comment by Zhengqbbb** [comment]
> "commitlint"

use husky git hook
https://github.com/Zhengqbbb/cz-git/blob/main/.husky/commit-msg

>  test them

test unitest or e2e test or just commit test ?
- unitest and e2e test can use github action: https://github.com/Zhengqbbb/cz-git/blob/main/.github/workflows/nodejs.yml
- commit msg test use commitlint with `husky`

> after merging changes release

Release is not automatic, it can be set to manual, using github action. Write a command that triggers running `npm release`
